custom_components/zaptec/api.py

```diff
@@ -380,33 +409,59 @@ async def _stream(self, cb=None, ssl_context=None):
             _LOGGER.exception("Stream failed: %s", err)
 
         finally:
-            # To ensure its not set if not active
+            # Cleanup
             self._stream_receiver = None
+            self._stream_running = False
+            _LOGGER.info("Servicebus stream stopped for %s", self.qual_id)
+
+    async def stream_close(self):
+        """Close the stream receiver."""
+        from azure.servicebus.exceptions import ServiceBusError
 
-    async def cancel_stream(self):
         try:
-            from azure.servicebus.exceptions import ServiceBusError
-        except ImportError:
-            return
+            if self._stream_receiver is not None:
+                await self._stream_receiver.close()
+        except ServiceBusError:
+            # This happens if the receiver is in the process of setting up
+            # or closing when are trying to close it.
+            pass
 
+    async def cancel_stream(self):
         if self._stream_task is not None:
+            await self.stream_close()
+            self._stream_task.cancel()
             try:
-                if self._stream_receiver is not None:
-                    await self._stream_receiver.close()
-                self._stream_task.cancel()
                 await self._stream_task
-                _LOGGER.debug("Canceled stream")
-            except (ServiceBusError, CancelledError):
+            except asyncio.CancelledError:
                 pass
-                # this will still raise a exception, I think its a 3.7 issue.
-                # recheck this when the i have updated to 3.9
-
             finally:
                 self._stream_task = None
 
-    # -----------------
-    # API methods
-    # -----------------
+    def stream_update(self, data: TDict):
+        """update for the stream."""
+
+        charger_id = data.pop("ChargerId", None)
+        if charger_id == "00000000-0000-0000-0000-000000000000":
+            _LOGGER.debug("Ignoring charger with id %s", charger_id)
+            return
+        elif charger_id is None:
+            _LOGGER.warning("Unknown update message %s", data)
+            return
+
+        if (charger := self._account.map.get(charger_id)) is None:
+            _LOGGER.warning("Got update for unknown charger id %s", charger_id)
+            return
+
+        charger = self.map.get(charger_id)
+        if charger is None:
+            _LOGGER.warning("Got update for unknown charger id %s", charger_id)
+            return
```

Comment:
This looks like a copypasta gone wrong, I think the first charger assignment is the correct one, and the second is an artifact from before the function was moved from the account object?